src/upgrades/pre_0_10_0/cas_migration.rs

```diff
@@ -0,0 +1,246 @@
+    ca_objects_migration: CaObjectsMigration,
+}
+
+impl CasStoreMigration {
+    pub fn prepare(mode: UpgradeMode, config: &Config) -> UpgradeResult<()> {
+        let current_kv_store = KeyValueStore::disk(&config.data_dir, CASERVER_DIR)?;
+        let new_kv_store = KeyValueStore::disk(&config.upgrade_data_dir(), CASERVER_DIR)?;
+        let new_agg_store = AggregateStore::<CertAuth>::disk(&config.upgrade_data_dir(), CASERVER_DIR)?;
+        let ca_objects_migration = CaObjectsMigration::create(config)?;
+
+        CasStoreMigration {
+            current_kv_store,
+            new_kv_store,
+            new_agg_store,
+            ca_objects_migration,
+        }
+        .prepare_new_data(mode)
+    }
+}
+
+impl UpgradeStore for CasStoreMigration {
+    fn needs_migrate(&self) -> Result<bool, PrepareUpgradeError> {
+        Ok(self.current_kv_store.version_is_after(KrillVersion::release(0, 9, 0))?
+            && self
+                .current_kv_store
+                .version_is_before(KrillVersion::candidate(0, 10, 0, 1))?)
+    }
+
+    fn prepare_new_data(&self, mode: UpgradeMode) -> Result<(), PrepareUpgradeError> {
+        // check existing version, wipe if needed
+        self.preparation_store_prepare()?;
+
+        info!(
+            "Prepare upgrading CA command and event data to Krill version {}",
+            KRILL_VERSION
+        );
+
+        // Migrate the event sourced data for each CA and create new snapshots
+        for scope in self.current_kv_store.scopes()? {
+            // Getting the Handle should never fail, but if it does then we should bail out asap.
+            let handle = CaHandle::from_str(&scope)
+                .map_err(|_| PrepareUpgradeError::Custom(format!("Found invalid CA handle '{}'", scope)))?;
+
+            // Get the info from the current store to see where we are
+            let mut data_upgrade_info = self.data_upgrade_info(&scope)?;
+
+            // Migrate the initialisation event, if not done in a previous run. This
+            // is a special event that has no command, so we need to do this separately.
+            if data_upgrade_info.last_event == 0 {
+                // Make a new init event.
+                let init_key = Self::event_key(&scope, 0);
+                let old_init: OldCaIni = self.get(&init_key)?;
+                let (id, _, old_ini_det) = old_init.unpack();
+                let ini = IniDet::new(&id, old_ini_det.into());
+                self.new_kv_store.store(&init_key, &ini)?;
+            }
+
+            // Get the list of commands to prepare, starting with the last_command we got to (may be 0)
+            let old_cmd_keys = self.command_keys(&scope, data_upgrade_info.last_command)?;
+
+            // Report the amount of (remaining) work
+            let total_commands = old_cmd_keys.len();
+            if data_upgrade_info.last_command == 0 {
+                info!("Will migrate {} commands for Publication Server", total_commands);
```

Comment:
Why does this say Publication Server when we are in CA migration code?

Reply:
oops.. fixed